Added support for multiple homes in pam_zfs_key module

dvaerum opened the pull request:
**Disclamer:** I don't normally code C, so this is a best-effort attempt. There can easily be something I have overlooked when it comes to memory handling. So I am open for feedback and help to improve the code 😁

This implemented support for having multiple datasets unlocked and mounted when a session is opened.
Example: `homes=rpool/home,tank/users`

Extra unit tests have been added

A man page document has been added `man 8 pam_zfs_key` (since none existed 😭). A few references to the new man page have also been added in other documents.

<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

### Motivation and Context
I have 2 pools on my system, both containing user data, so I want to be able to easily decrypt and mount the user dataset on both pools on user login.
So I have made this patch for the pam_zfs_key module, which worked on my system. So I thought I wanted to share it.
I want to upstream this patch because there may be other people with this problem.

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

### Description
This patch adds support for comma-separated homes= prefixes in pam_zfs_key and makes it so that pam_zfs_key can manage encrypted ZFS datasets across multiple pools. This patch is, as far as I can tell, fully backward compatible with existing PAM configs.

### How Has This Been Tested?
I created a VM setup for the development and testing. When I applied the patch on my workstation. For the final test.
I am running NixOS, so I simply applied the patch as an overlay and rebuilt my system.

```nix
nixpkgs.overlays = [
  ( final: prev:  let
      zfs-patch = prev.fetchpatch {
        url = "https://github.com/dvaerum/zfs/commit/7ae5eb5797a5309f943601453a871460b1caefe2.patch";
        hash = "sha256-QGoR3S2vWbaxoFTqnKjgdyKkI3juKTsfO98UU2rEfi4=";
      };
    in {
      zfs = prev.zfs.overrideAttrs (old: {
        patches = (old.patches or [ ]) ++ [ zfs-patch ];
      });
      zfs_2_4 = prev.zfs_2_4.overrideAttrs (old: {
        patches = (old.patches or [ ]) ++ [ zfs-patch ];
      });
      zfs_unstable = prev.zfs_unstable.overrideAttrs (old: {
        patches = (old.patches or [ ]) ++ [ zfs-patch ];
      });
    }
  )
];

```

I have expanded the existing unittests for the PAM module to also cover multi-homes config to make sure that this feature continues to work, if accepted. However, I may not have covered everything, but if you see missing test scenarios, I will try to add them 😁 

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Quality assurance (non-breaking change which makes the code more robust against bugs)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

tonyhutter commented:
I don't use encryption so maybe this is a dumb question - can `homes` be any dataset, or does it have to be home directories specifically?  If it can be any dataset, would it make sense to name it something more generic, like just `datasets`?

tonyhutter commented:
> Another reason for me to keep homes= was to keep backward-compatibility with people's config.

Ok, if people have already been using it as `homes`, then that's good enough reason to keep the name as-is.

dvaerum commented:
> > > Is there a danger in going over 16 entries?
> 
> > Not that I know of, and I don't know what that would be 😅. I was just not really sure that allowing "infinity" was a good idea either
> 
> I would just remove the limit. We should assume that the `config->homes_prefix` being passed in is valid.

Okay, I have removed the limit and added your unittests 😁

tonyhutter commented:
Looks like some of the tests aren't passing now:
```
    FAIL pam/pam_basic (expected PASS)
    FAIL pam/pam_change_unmounted (expected PASS)
    FAIL pam/pam_nounmount (expected PASS)
    FAIL pam/pam_short_password (expected PASS)
```

dvaerum commented:
@tonyhutter the tests are passing again. I just had to add support index argument to the keystatus_mh function, so that it worked with the unittest I got from you  😁

**Before:**
```bash
function keystatus_mh {
    log_must [ "$(get_prop keystatus "$TESTPOOL/pam-multi-home/${username}")" = "$1" ]
}
```

**After:**
```bash
function keystatus_mh {
    typeset suffix="${2:-}"
    log_must [ "$(get_prop keystatus "$TESTPOOL/pam-multi-home${suffix}/${username}")" = "$1" ]
}
```

dvaerum commented:
Thanks @tonyhutter I am looking forward to seeing it in upstream 😁
